### DISCUSSION TITLE
Question about using Swift DocC archive with Docusaurus

### DISCUSSION THREAD

**Discussion by jkboxomine** [facebook/docusaurus#7839] [Q&A]
We're trying to use a [Swift-DocC](https://github.com/apple/swift-docc) output as part of our static site generator built on Docusaurus.

The site is internationalized in English and one other language, but the API documentation is in English only. We're simply trying to show the English API documentation regardless of the language setting of the website.

We have a single DocC output placed under the `/static` directory, in a similar manner to API documentation for other platforms (generated by [Dokka](https://github.com/Kotlin/dokka), [TypeDoc](https://typedoc.org/), etc.), and the link to the documentation is configured with `pathname:///` protocol on the sidebar file.

The content of DocC output is displayed properly in English mode. However, when we switch to the other language, the following occurs (Updated: 2022-08-11)

* In the local preview, a blank screen shows up with `Uncaught SyntaxError: Unexpected token <` error on the browser console.
* In the GitHub Pages hosting, "The page you’re looking for can’t be found." shows up.

When the language is switched, the base path is changed on the URL, from `{docs.domain.com}/{sub-path}` (English) to `{docs.domain.com}/{lang_code}/{sub-path}` (the other language), as expected. For the API documentation of other platforms, there's no issue with the display of content. For the DocC output, however, the issue described above occurs.

We think it might be due to the conflict between [routing behavior](https://docusaurus.io/docs/next/advanced/routing) of Docusaurus and DocC archive, but could not find an exact cause or workaround yet.

- If anyone succeeded in using Swift-DocC output with Docusaurus (i18n enabled), we'd appreciate it if you could share your experience or thoughts.
- As a workaround, if there's a way to force a specific link to be routed to the URL without `lang_code` when the website is set to non-English mode, let us know.

**Comment by slorber** [comment]
Hey

It's not very clear to me how you integrated Swift-DocC

Can you show me your live site, eventually source code? Because not seeing the problem myself makes it hard to debug, and I can't spend hours trying to figure out how you did encounter this problem in the first place.

What prevents you from creating a smaller repro with some fake HTML files instead of the Swift-DocC output? Is this issue really related to DocC in the first place?

**Comment by jkboxomine** [comment]
> Is this issue really related to DocC in the first place?

Yes. Outputs from other API generation tools do not have this issue, which allow specifying `index.html` at the end of `pathname:///`. With DocC output, specifying `index.html` does not work due to some peculiarity, which is why [`sidebar.js`](https://github.com/jkboxomine/dcsrsTest/blob/main/sidebars.js) should be configured as such.

I updated the issue description on the original posting, and created a test site and a GitHub repo for the reproduction of this issue. You can test the issue behavior by clicking "API Reference" link on the sidebar with the language set to English or French.

* Test page on GitHub Pages: https://jkboxomine.github.io/dcsrsTest/docs/intro
* GitHub repo: https://github.com/jkboxomine/dcsrsTest

I also reported this to [Swift forum](https://forums.swift.org/t/issue-using-docc-archive-with-react-based-static-site-generators-routing/59072/5). If the issue cannot be mitigated on the Swift DocC side, any workaround on Docusaurus would be helpful.

**Comment by jkboxomine** [comment]
@slorber I received an [answer](https://github.com/apple/swift-docc-plugin#previewing-documentation) from the maintainer of [Swift-DocC plugin](https://github.com/apple/swift-docc-plugin), which is the tool to transform the DocC Archive into a format for static hosting. He says we have to generate a separate set of documentation with a different base path configuration (including `/lang_code` in this case). I believe it would require us to place this in a different directory from the one for English, under `static`.

As far as I can tell, using a sidebar item with `type: 'link'` and `pathname:///` protocol expects using a single instance of file set under `static`. If this is true, the current behavior of Swift-DocC plugin conflicts with the Docusaurus in terms of i18n requirement, making the Swift-DocC plugin output not usable on multi-lingual Docusarus sites.

I would appreciate it if you could confirm this or provide some other suggestion or workaround.

**Comment by jkboxomine** [comment]
@slorber In any case, we are aware that the root cause of the issue lies in the Swift-DocC plugin (and DocC archive), since its output is not a pure static HTML like the output of other API documentation generators. We just need evidence from both sides to prove that we must switch to an alternative method for iOS API documentation as long as we use Docusaurus and i18n.

**Comment by slorber** [comment]
hey @jkboxomine , this issue is extremely hard for me to decipher 😅 I'm not even sure I should read your wall of text because your analysis of the problem seems wrong to me.

Please give me a step by step guide to understand your problem. Make it crystal clear, no ambiguity, no possible wrong interpretation from my side, and eventually include screenshots.


---

I'm not 100% sure but what I think the problem is:

You have this sidebar external link:

```js
    {
      type: 'link',
      label: 'API Reference',
      href: 'pathname:///api-reference/documentation/deckofplayingcards',
    },
```

- Step1: open https://jkboxomine.github.io/dcsrsTest/fr/docs/intro (french doc)
- Step2: click the "API reference" external link in the sidebar

![CleanShot 2022-08-12 at 11 06 26@2x](https://user-images.githubusercontent.com/749374/184322681-639a40bf-9a27-4b5d-b901-e803fd34d76a.png)

Result:

https://jkboxomine.github.io/dcsrsTest/fr/api-reference/documentation/deckofplayingcards

<img width="806" alt="CleanShot 2022-08-12 at 11 09 10@2x" src="https://user-images.githubusercontent.com/749374/184323050-a14e1bb6-4b2d-4a57-9831-84f3ea7c983e.png">


Expected:

https://jkboxomine.github.io/dcsrsTest/api-reference/documentation/deckofplayingcards

<img width="1134" alt="CleanShot 2022-08-12 at 11 09 40@2x" src="https://user-images.githubusercontent.com/749374/184323119-0c74ae66-709c-47e5-aa13-8dfa7cd75e26.png">


---

Is that the problem you are reporting here?

If it's not the problem you are reporting: please provide me crystal clear reproduction instructions

---

> For clarification, the API documentation itself is not internationalized, that is, in English only. We're just trying to show the English API documentation whether the website is set in English or in the other language. 

If that's the case, you should decide on a unique URL for this English API doc to be hosted on. I believe you don't want/need this API doc to be accessible on multiple different URLs, and there's no value to make it available under a `/fr/` baseurl => this would just be a duplicate.

> But currently it fails when the website is set to the other language.

To me, this is a linking problem: the link rendered by Docusaurus on the French site is simply not what you want it to be: it shouldn't contain the `/fr/` prefix for the French site. This has nothing to do with the Docusaurus / Swift-DocC integration, it is just not the external link you want.

This is not implemented yet, but we should enable you do solve this problem with some config:

```js
    {
      type: 'link',
      label: 'API Reference',
      href: 'pathname:///api-reference/documentation/deckofplayingcards',
      autoAddBaseUrl: false,
    },
```

**Comment by jkboxomine** [comment]
@slorber Thanks for taking your time to summarize the issue. Your interpretation is correct and your suggested solution should work. What do I need to do to request adding the config described above?

**Reply by slorber** [reply]
It will be solved in https://github.com/facebook/docusaurus/pull/7949, and once it's merged you have to wait for 2.1 release or use a canary version

**Reply by jkboxomine** [reply]
Thank you for handling the request so promptly.
